core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -1250,6 +1362,147 @@ public boolean move() {
+
+        @Nullable
+        private final Decision canRebalanceDecision;
+        @Nullable
+        private final Map<String, NodeRebalanceDecision> nodeDecisions;
+        private float currentWeight;
+
+        protected RebalanceDecision(Decision canRebalanceDecision, Type finalDecision, String finalExplanation) {
+            this(canRebalanceDecision, finalDecision, finalExplanation, null, null, Float.POSITIVE_INFINITY);
+        }
+
+        protected RebalanceDecision(Decision canRebalanceDecision, Type finalDecision, String finalExplanation,
+                                    String assignedNodeId, Map<String, NodeRebalanceDecision> nodeDecisions, float currentWeight) {
+            super(finalDecision, finalExplanation, assignedNodeId);
+            this.canRebalanceDecision = canRebalanceDecision;
+            this.nodeDecisions = nodeDecisions != null ? Collections.unmodifiableMap(nodeDecisions) : null;
+            this.currentWeight = currentWeight;
+        }
+
+        /**
+         * Creates a new {@link RebalanceDecision}, computing the explanation based on the decision parameters.
+         */
+        public static RebalanceDecision decision(Decision canRebalanceDecision, Type finalDecision, String assignedNodeId,
+                                                 Map<String, NodeRebalanceDecision> nodeDecisions, float currentWeight, float threshold) {
+            final String explanation = produceFinalExplanation(finalDecision, assignedNodeId, threshold);
+            return new RebalanceDecision(canRebalanceDecision, finalDecision, explanation, assignedNodeId, nodeDecisions, currentWeight);
+        }
+
+        /**
+         * Returns the decision for being allowed to rebalance the shard.
+         */
+        @Nullable
+        public Decision getCanRebalanceDecision() {
+            return canRebalanceDecision;
+        }
+
+        /**
+         * Gets the individual node-level decisions that went into making the final decision as represented by
+         * {@link #getFinalDecisionType()}.  The map that is returned has the node id as the key and a {@link NodeRebalanceDecision}.
+         */
+        @Nullable
+        public Map<String, NodeRebalanceDecision> getNodeDecisions() {
+            return nodeDecisions;
+        }
+
+        private static String produceFinalExplanation(final Type finalDecisionType, final String assignedNodeId, final float threshold) {
+            final String finalExplanation;
+            if (assignedNodeId != null) {
+                if (finalDecisionType == Type.THROTTLE) {
+                    finalExplanation = "throttle moving shard to node [" + assignedNodeId + "], as it is " +
+                                           "currently busy with other shard relocations";
+                } else {
+                    finalExplanation = "moving shard to node [" + assignedNodeId + "] to form a more balanced cluster";
```

Comment:
It could totally be subsequent work, but I think it'd be great if we could expose the deltas in the explanation also (I mean, for us diagnosing allocation issues it'd be nice to have)